gossip/gossip_test.go

```diff
@@ -0,0 +1,151 @@
+	gossiper := NewGossiper[testTx](config, nil, nil, nil, 0)
+	done := make(chan struct{})
+	wg := &sync.WaitGroup{}
+
+	wg.Add(1)
+	go gossiper.Gossip(done, wg)
+
+	close(done)
+	wg.Wait()
+}
+
+func TestGossiperGossip(t *testing.T) {
+	tests := []struct {
+		name      string
+		requester []*testTx // what we have
+		responder []*testTx // what the peer we're requesting gossip from has
+		expected  []*testTx // what we should have after a gossip cycle
+	}{
+		{
+			name: "no gossip - no one knows anything",
+		},
+		{
+			name:      "no gossip - requester knows more than responder",
+			requester: []*testTx{{hash: Hash{0}}},
+			expected:  []*testTx{{hash: Hash{0}}},
+		},
+		{
+			name:      "no gossip - requester knows everything responder knows",
+			requester: []*testTx{{hash: Hash{0}}},
+			responder: []*testTx{{hash: Hash{0}}},
+			expected:  []*testTx{{hash: Hash{0}}},
+		},
+		{
+			name:      "gossip - requester knows nothing",
+			responder: []*testTx{{hash: Hash{0}}},
+			expected:  []*testTx{{hash: Hash{0}}},
+		},
+		{
+			name:      "gossip - requester knows less than responder",
+			requester: []*testTx{{hash: Hash{0}}},
+			responder: []*testTx{{hash: Hash{0}}, {hash: Hash{1}}},
+			expected:  []*testTx{{hash: Hash{0}}, {hash: Hash{1}}},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+			ctrl := gomock.NewController(t)
+
+			cc := codec.NewManager(units.MiB)
+			lc := linearcodec.NewDefault()
+			require.NoError(lc.RegisterType(PullGossipRequest{}))
+			require.NoError(lc.RegisterType(PullGossipResponse{}))
+			require.NoError(cc.RegisterCodec(0, lc))
+
+			responseSender := common.NewMockSender(ctrl)
+			responseRouter := p2p.NewRouter(logging.NoLog{}, responseSender)
+			responseBloom, err := NewBloomFilter(1000, 0.01)
+			require.NoError(err)
+			responseSet := testSet{
+				set:   set.Set[*testTx]{},
+				bloom: responseBloom,
+			}
+			for _, item := range tt.responder {
+				require.NoError(responseSet.Add(item))
+			}
+			handler := NewHandler[*testTx](responseSet, cc, 0)
+			_, err = responseRouter.RegisterAppProtocol(0x0, handler)
+			require.NoError(err)
+
+			requestSender := common.NewMockSender(ctrl)
+			requestRouter := p2p.NewRouter(logging.NoLog{}, requestSender)
+			require.NoError(requestRouter.Connected(context.Background(), ids.EmptyNodeID, nil))
+
+			gossiped := make(chan struct{})
+			requestSender.EXPECT().SendAppRequest(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+				Do(func(ctx context.Context, nodeIDs set.Set[ids.NodeID], requestID uint32, request []byte) {
+					go func() {
+						require.NoError(responseRouter.AppRequest(ctx, ids.EmptyNodeID, requestID, time.Time{}, request))
+					}()
+				}).AnyTimes()
+
+			responseSender.EXPECT().
+				SendAppResponse(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+				Do(func(ctx context.Context, nodeID ids.NodeID, requestID uint32, appResponseBytes []byte) {
+					go func() {
```

Comment:
nit: I don't think the response handler needs to be on a goroutine

Reply:
We need to make this call in a goroutine so the client lock can be released. The client internally holds a lock on the requestID map that's only released once the `SendAppRequest` call completes. If we don't make this async, the client is trying to get a response, but it can't because it's blocked on the original sending thread still being blocked.

Reply:
This shouldn't happen in real-life because sending over the network is actually async, this is a unit-test hack to simulate sending a message over the wire.